language/move-prover/bytecode-to-boogie/src/bytecode_instrs.bpl

```diff
@@ -15,16 +18,76 @@ function {:constructor} Path(p: [int]Edge, size: int): Path;
+function {:constructor} TypeValueArray(v: [int]TypeValue, l: int): TypeValueArray;
+
+const EmptyTypeValueArray: TypeValueArray;
+axiom l#TypeValueArray(EmptyTypeValueArray) == 0;
+axiom (forall v1, v2: [int]TypeValue, l1, l2: int ::
+         TypeValueArray(v1, l1) == TypeValueArray(v2, l2) <==>
+         l1 == l2 && (forall i: int :: i >= 0 && i < l1 ==> v1[i] == v2[i]));
+
 type {:datatype} Value;
+type {:datatype} ValueArray;
 function {:constructor} Boolean(b: bool): Value;
 function {:constructor} Integer(i: int): Value;
 function {:constructor} Address(a: Address): Value;
 function {:constructor} ByteArray(b: ByteArray): Value;
 function {:constructor} Str(a: String): Value;
-function {:constructor} Map(m: [Edge]Value): Value;
-function {:constructor} Vector(v: [Edge]Value, l: int): Value;
-
-const DefaultMap: [Edge]Value;
+function {:constructor} Vector(v: ValueArray): Value;
+function {:constructor} Struct(v: ValueArray): Value;
+function {:constructor} ValueArray(v: [int]Value, l: int): ValueArray;
+
+function vlen(v: Value): int { l#ValueArray(v#Vector(v)) }
+function vmap(v: Value): [int]Value { v#ValueArray(v#Vector(v)) }
+function mk_vector(v: [int]Value, l: int): Value { Vector(ValueArray(v, l)) }
+function slen(v: Value): int { l#ValueArray(v#Struct(v)) }
+function smap(v: Value): [int]Value { v#ValueArray(v#Struct(v)) }
+function mk_struct(v: [int]Value, l: int): Value { Struct(ValueArray(v, l)) }
+
+
+const EmptyValueArray: ValueArray;
+axiom l#ValueArray(EmptyValueArray) == 0;
+axiom (forall v1, v2: [int]Value, l1, l2: int ::
+         ValueArray(v1, l1) == ValueArray(v2, l2) <==>
+         l1 == l2 && (forall i: int :: i >= 0 && i < l1 ==> v1[i] == v2[i]));
+
+// Axiomatic definition of type test and equality.
+function get_type(v: Value): TypeValue;
+function has_type(t: TypeValue, v: Value): bool;
+function is_equal(t: TypeValue, v1: Value, v2: Value): bool;
+
+axiom (forall x: bool :: get_type(Boolean(x)) == BooleanType());
+axiom (forall x: int :: get_type(Integer(x)) == IntegerType());
+axiom (forall x: Address :: get_type(Address(x)) == AddressType());
+axiom (forall x: ByteArray :: get_type(ByteArray(x)) == ByteArrayType());
+axiom (forall x: String :: get_type(Str(x)) == StrType());
+axiom (forall va: ValueArray :: (exists tv: TypeValue ::
+        (l#ValueArray(va) > 0 ==> tv == get_type(v#ValueArray(va)[0]))
+        && get_type(Vector(va)) == VectorType(tv)));
+axiom (forall va: ValueArray :: (exists tn: TypeName, tva: TypeValueArray ::
+        l#ValueArray(va) == l#TypeValueArray(tva)
+        && (forall i: int :: i >= 0 && i < l#ValueArray(va) ==>
+              v#TypeValueArray(tva)[i] == get_type(v#ValueArray(va)[i]))
+        && get_type(Struct(va)) == StructType(tn, tva)));
+
+axiom (forall v: Value, tv: TypeValue :: has_type(tv, v) <==> get_type(v) == tv);
+
+axiom (forall tv: TypeValue, v1, v2: Value :: is_equal(tv, v1, v2) <==> v1 == v2);
```

Comment:
Let us make is_equal and inline function for now.

Reply:
Done